src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -1660,4 +1736,72 @@ public int doFinal(ByteBuffer src, ByteBuffer dst) {
             return l;
         }
     }
+
+    /**
+     * This class is for decryption when both GCTR and GHASH
+     * can operation in parallel.
+     */
+    static final class DecryptOp implements GCMOperation {
+        GCTR gctr;
+        GHASH ghash;
+
+        DecryptOp(GCTR c, GHASH g) {
+            gctr = c;
+            ghash = g;
+        }
+
+        @Override
+        public int update(byte[] in, int inOfs, int inLen, byte[] out,
+            int outOfs) {
+            ghash.update(in, inOfs, inLen);
+            return gctr.update(in, inOfs, inLen, out, outOfs);
+        }
+
+        @Override
+        public int update(byte[] in, int inOfs, int inLen, ByteBuffer dst) {
+            ghash.update(in, inOfs, inLen);
+            return gctr.update(in, inOfs, inLen, dst);
+        }
+
+        @Override
+        public int update(ByteBuffer src, ByteBuffer dst) {
+            src.mark();
+            ghash.update(src, src.remaining());
+            src.reset();
+            return gctr.update(src, dst);
+        }
+
+        @Override
+        public int doFinal(byte[] in, int inOfs, int inLen, byte[] out,
+                           int outOfs) {
+            int len = 0;
+            if (inLen >= PARALLEL_LEN) {
+                implGCMCrypt(in, inOfs, inLen, in, inOfs, out, outOfs, gctr,
```

Comment:
Should save the return value into 'len'?

Reply:
I do not understand this comment

Reply:
Doesn't implGCMCrypt(...) return an int telling how much bytes it has processed? Then we adjust the index and remain input length with this value. But here we didn't save the return value which looks wrong. Did I miss something?

Never mind my second comment, I mis-read the code.

Reply:
Ah.. I see.. yes, it should be included in len